hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/capacity/LeafQueue.java

```diff
@@ -2330,25 +2330,27 @@ void updateMaximumApplications(CapacitySchedulerConfiguration conf) {
     int maxAppsForQueue = conf.getMaximumApplicationsPerQueue(getQueuePath());
 
     int maxGlobalApplications = conf.getGlobalMaximumApplicationsPerQueue();
-    if (maxGlobalApplications < 0) {
-      maxGlobalApplications = conf.getMaximumSystemApplications();
-    }
+    int maxSystemApplications = conf.getMaximumSystemApplications();
+    int baseMaxApplications = maxGlobalApplications > 0 ?
```

Comment:
It is hard to see how absolute config type is handled and when the maxSystemApplications is really used.

In the else branch (line 2342) we either have maxGlobalApplications <= 0 or Absolute capacity config type, maxSystemApplications is used in the first case maxGlobalApplications used in the later case.


Maybe something like this would be easier to read:
```
...
If (this.capacityConfigType == CapacityConfigType.ABSOLUTE_RESOURCE) {
    maxLabel, maxAppsForQueue = getUpdatedMaxAppsForQueueFromLabels(maxGlobalApplications);
} else {
  If (maxGlobalApplications > 0) {
    maxAppsForQueue = maxGlobalApplications; 
  } else {
    maxLabel, maxAppsForQueue = getUpdatedMaxAppsForQueueFromLabels(maxSystemApplications);
  }
}
...


ImmutablePair<String, Integer> getUpdatedMaxAppsForQueueFromLabels(int base) {
    String maxLabel = RMNodeLabelsManager.NO_LABEL;
    int maxAppsForQueue = conf.getMaximumApplicationsPerQueue(getQueuePath());
    for (String label : configuredNodeLabels) {
        int maxApplicationsByLabel = (int) (base  * queueCapacities.getAbsoluteCapacity(label));
        if (maxApplicationsByLabel > maxAppsForQueue) {
             maxAppsForQueue = maxApplicationsByLabel;
             maxLabel = label;
        }
   }
  return new ImmutablePair<>(maxLabel, maxAppsForQueue);
}
```

It would be much simpler, but the maxLabel is needed for the log message later.

Reply:
The problem with this approach is that if using ABSOLUTE_RESOURCE, the maxGlobalApplications is used, regardless of the fact that it could be -1. You would need yet an other check there, which would make this logic even more convoluted. In addition to this, using pairs in Java is verbose because it lacks destructuring, so it would look more like:
```java
ImmutablePair<String, Integer> maxLabelAndApp = getUpdatedMaxApps(int base);
maxLabel = maxLabelAndApp.get(0);
maxAppsForQueue = maxLabelAndApp.get(1);
```
In my honest opinion, this whole logic should be plain simple, but there are already tests that build on this behaviour, not to mention users, therefore we need to be backward compatible.

Reply:
OK, thanks for the clarification.